src/main/java/io/jenkins/plugins/ml/IPythonBuilder.java

```diff
@@ -0,0 +1,132 @@
+/*
+ * The MIT License
+ *
+ * Copyright 2020 Loghi Perinpanayagam.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package io.jenkins.plugins.ml;
+
+import hudson.Extension;
+import hudson.Launcher;
+import hudson.model.AbstractBuild;
+import hudson.model.AbstractProject;
+import hudson.model.BuildListener;
+import hudson.tasks.BuildStepDescriptor;
+import hudson.tasks.Builder;
+import hudson.util.FormValidation;
+import org.jenkinsci.Symbol;
+import org.kohsuke.stapler.DataBoundConstructor;
+import org.kohsuke.stapler.QueryParameter;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Hashtable;
+import java.util.Map;
+
+public class IPythonBuilder extends Builder {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(IPythonBuilder.class);
+    private static final Map<IPythonUserConfig,IPythonInterpreterManager> configBasedManagerMap = new Hashtable<>();
```

Comment:
I think we can't use this approach @Loghijiaha .

I will have time to debug it over the next week, but if you have not debugged Jenkins plug-ins, this is a great chance.

1. In my case, I use Eclipse, but IntelliJ should work too; start your plug-in with the debugger. In Eclipse that's Debug As, choose Maven, `hpi:run` target (making sure sources are properly mapped)
2. Visit the page where you validate the builder configuration, and trigger the validation 10 times, every time with changes in the configuration (just increase the `int`, for instance)
3. Now each `IPythonUserConfig` should return `false` in the `equals()`, and cause a new entry to be added to this `Map`
4. Set a break-point now in your `doValidate` method
5. Trigger the validation once again

I haven't tested the code, so I could be wrong, but we would need to answer the following questions:

1. How many objects you have in this `Map`?
2. What happens if we get 9_999_999_999 requests with slightly different `IPythonUserConfig`'s?

I think `Map`'s are one of the most common sources of memory leaks in Java code, so better make sure we don't introduce one unless necessary.

IMO, we can create an interpreter to validate, and discard it. And again, we would have to think where/when to initialize the interpreter then.

Cheers
Bruno

Reply:
Agree on that @bruno. If the user already configured with the same server, it will return the `interpreterManager ` respect to the server configuration by checking the existence of the server configuration. If it is null, it will create a new interpreter manager and add to the HashTable. In worst case, we can have 9_999_999_999 requests. If the user needs to do with 9_999_999_999 different server it is up to the user. If some of the requests are invoked using same configuration it will use an existing one instead of creating a new one.
When a server needs to be created while changing one of these parameters
- server address
- launch timeout
- max results

```
IPythonInterpreterManager interpreterManager = configBasedManagerMap.get(jobUserConfig) ;

        if(interpreterManager == null ){
            interpreterManager = new IPythonInterpreterManager(jobUserConfig);
            interpreterManager.initiateInterpreter();

```
I am debugging every object in the plugin. We will discuss in the next meeting.
Cheers :)